[MemCpyOpt] fix incorrect handling of lifetime markers

vtjnash opened the pull request:
Having lifetime markers should only increase the information available
to LLVM, but it would instead rely on the callback to entirely give up
if it encountered a lifetime marker that wasn't full size, but
sub-optimal lifetime markers are not supposed to be forbidding
optimizations that would otherwise apply if they were either absent or
optimal. This pass wasn't tracking GEP offsets either, so it wasn't
quite correctly handled either, although earlier sub-optimal checks
that this size is the same as the alloca test made this safe in the
past, and unlikely to have encountered anything else in the past.

~~Having lifetime markers should only increase the information available to LLVM, but it was stopping the MemSSA walk and unable to see that these were extraneous. Refactored `hadUndefContentsBefore` to try to DRY the code slightly, so that we are doing that walk inside it, instead of in every caller.~~

~~Secondly, it would ignore a lifetime marker that wasn't full size (relying on the callback to forbid the optimization entirely), but again sub-optimal lifetime markers are not supposed to be forbidding optimizations that would otherwise apply if they were either absent or optimal. This pass wasn't tracking GEP offsets either, so it wasn't quite correctly handled either (although earlier sub-optimal checks that this size is the same as the alloca test made this safe in the past).~~

~~Lastly, the stack-move test seemed to have been providing lifetime in bits instead of bytes, although now we optimize it regardless of that, so we didn't really have to fix that.~~

vtjnash commented:
This is a dependent commit of https://github.com/vtjnash/llvm-project/commit/645f2472085420b9e9a4e4c1b59c4c1768a31e28, which removes the function's requirement of having this constant size be computed and available.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1820,7 +1835,8 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
 
-    if (hasUndefContents(MSSA, BAA, M->getSource(), MD, M->getLength())) {
+    if (hadUndefContentsBefore(MSSA, BAA, M->getSource(), AnyClobber, SrcLoc,
+                               M->getLength())) {
```
Isn't this going to recompute the clobbering access now, even though we already know it?

vtjnash replied:
Yes. I assumed it was inexpensive since it is a caching walker, but I could add a bool parameter if necessary to optimize that

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1366,56 +1366,68 @@ bool MemCpyOptPass::processMemSetMemCpyDependence(MemCpyInst *MemCpy,
+        continue;
+      } else if (II->getIntrinsicID() == Intrinsic::lifetime_end) {
+        Clobber = Def->getDefiningAccess();
+        continue;
```
Lack of break below here is bypassing the MSSA clobber walk limit.

vtjnash replied:
There is a `return false` immediately below here

nikic replied:
Oh yeah, there is. I was confused by the diff rendering. In that case I'm not sure what is going on.

vtjnash replied:
I didn't read the tests closely, so I didn't realize those were often reads from undef it was removing. I thought that might have been GVN's job to detect a memcpy from undef contents and remove it?

I see at least one place there that mentions it shouldn't fail to optimize when encountering lifetime there:
https://github.com/llvm/llvm-project/blob/ce747a16328b2fbc365e1cb1cb01cb400c2c1b4c/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L4916-L4925

vtjnash replied:
I've trimmed this PR down to just the part I needed for https://github.com/vtjnash/llvm-project/commit/645f2472085420b9e9a4e4c1b59c4c1768a31e28, pending separate discussion of whether it is even worthwhile optimizing those tests for which lifetime.start does not cover the whole alloca.

Additionally, I made a separate PR as you suggested (https://github.com/llvm/llvm-project/pull/143958) to handle those simple cases in InstCombine instead, and to see how that affects the benchmarks.

nikic reviewed (changes requested):
In practice, all lifetime markers should have an alloca argument and a size matching the alloca and be handled by the "whole alloca" code path that is independent of the passed size. In theory, lifetimes can be used in other ways, but this is the only practically relevant case. If I ever find the time, I'll drop the size argument from these intrinsics.

I do see some diffs from this change on https://github.com/dtcxzyw/llvm-opt-benchmark/pull/2428, but I'm having a hard time understanding where they are actually coming from. None of the cases I looked at have mismatched lifetimes.

I think what actually happens is that your new implementation of hadUndefContentsBefore does not bail out if none of the cases match, which means that it will just continue the walk from the last found clobber -- which can only improve things if the clobber was the result of hitting the MSSA walk limit, rather than an actual clobber. So effectively, all this patch is really doing is add a bypass for the walk limit.

vtjnash commented on llvm/test/Analysis/ScopedNoAliasAA/alias-scope-merging.ll:
Thank you

vtjnash committed:
stop walk

vtjnash committed:
remove walk

vtjnash committed:
[memcpyopt] fix incorrect handling of lifetime markers

Having lifetime markers should only increase the information available
to LLVM, but it would instead rely on the callback to entirely give up
if it encountered a lifetime marker that wasn't full size, but
sub-optimal lifetime markers are not supposed to be forbidding
optimizations that would otherwise apply if they were either absent or
optimal. This pass wasn't tracking GEP offsets either, so it wasn't
quite correctly handled either, although earlier sub-optimal checks that
this size is the same as the alloca test made this safe in the past, and
unlikely to have encountered anything else in the past.

vtjnash committed:
Merge branch 'main' into jn/memcpyopt-lifetimes

vtjnash commented:
@nikic can you clear your request for changes, since I remove that part of the commit and just left the parts that will allow improving this code optimization in a stacked commit (https://github.com/vtjnash/llvm-project/commit/645f2472085420b9e9a4e4c1b59c4c1768a31e28), now that the InstCombine changes are done to hopefully avoid many of the nuisance benchmark changes originally caused here.

nikic commented:
Just like the previous changes, this still only improves handling for the case where the lifetime size does not match the alloca size, which should not occur in practice. Why do we need to handle it?

Note that this code originally preserved lifetime markers (roughly based on the NCD of the lifetimes) and that got "temporarily" dropped due to some issues and never restored. I don't think the current code cares about the size, but if we wanted to restore this (which we really should...) the size would matter again.

vtjnash commented:
bump?

nikic reviewed (approved):
LGTM with the additional comments you added removed. I've finally gotten around to removing the size argument from these intrinsics (https://github.com/llvm/llvm-project/pull/150248), so this is now just landing part of that early.

vtjnash committed:
Merge remote-tracking branch 'upstream/main' into jn/memcpyopt-lifetimes

vtjnash committed:
remove comment
